drivers/sensor/max17262/max17262.c

```diff
@@ -0,0 +1,345 @@
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialise the fuel gauge
+ *
+ * @param dev MAX17262 device to access
+ * @return 0 for success
+ * @return -EINVAL if the I2C controller could not be found
+ */
+static int max17262_gauge_init(const struct device *dev)
+{
+	struct max17262_data *priv = dev->data;
+	const struct max17262_config *const config = dev->config;
+	int16_t tmp, hibcfg;
+
+	priv->i2c = device_get_binding(config->bus_name);
+	if (!priv->i2c) {
+		LOG_ERR("Could not get pointer to %s device", config->bus_name);
+		return -EINVAL;
+	}
+
+	/* Read Status register */
+	max17262_reg_read(priv, STATUS, &tmp);
+
+	if (!(tmp & STATUS_POR)) {
+		/*
+		 * Status.POR bit is set to 1 when MAX17262 detects that
+		 * a software or hardware POR event has occurred and
+		 * therefore a custom configuration needs to be set...
+		 * If POR event did not happen (Status.POR == 0), skip
+		 * init and continue with measurements.
+		 */
+		LOG_DBG("No POR event detected - skip device configuration");
+		return 0;
+	}
+	LOG_DBG("POR detected, setting custom device configuration...");
+
+	/** STEP 1 */
+	max17262_reg_read(priv, FSTAT, &tmp);
+
+	/* Do not continue until FSTAT.DNR bit is cleared */
+	while (tmp & FSTAT_DNR) {
+		k_sleep(K_MSEC(10));
+		max17262_reg_read(priv, FSTAT, &tmp);
+	}
+
+	/** STEP 2 */
+	/* Store original HibCFG value */
+	max17262_reg_read(priv, HIBCFG, &hibcfg);
+
+	/* Exit Hibernate Mode step 1 */
+	max17262_reg_write(priv, SOFT_WAKEUP, 0x0090);
+	/* Exit Hibernate Mode step 2 */
+	max17262_reg_write(priv, HIBCFG, 0x0000);
+	/* Exit Hibernate Mode step 3 */
+	max17262_reg_write(priv, SOFT_WAKEUP, 0x0000);
+
+	/** STEP 2.1 --> OPTION 1 EZ Config (No INI file is needed) */
+	/* Write DesignCap */
+	max17262_reg_write(priv, DESIGN_CAP, config->design_cap);
+
+	/* Write IChgTerm */
+	max17262_reg_write(priv, ICHG_TERM, config->desired_charging_current);
+
+	/* Write VEmpty */
+	max17262_reg_write(priv, VEMPTY, ((config->empty_voltage / 10) << 7) |
+					  ((config->recovery_voltage / 40) & 0x7F));
+
+	/* Write ModelCFG */
+	if (config->charge_voltage > 4275)
+		max17262_reg_write(priv, MODELCFG, 0x8400);
+	else
+		max17262_reg_write(priv, MODELCFG, 0x8000);
+
+	/*
+	 * Read ModelCFG.Refresh (highest bit),
+	 * proceed to Step 3 when ModelCFG.Refresh == 0
+	 */
+	max17262_reg_read(priv, MODELCFG, &tmp);
+
+	/* Do not continue until ModelCFG.Refresh == 0 */
+	while (tmp & MODELCFG_REFRESH) {
+		k_sleep(K_MSEC(10));
+		max17262_reg_read(priv, MODELCFG, &tmp);
+	}
+
+	/* Restore Original HibCFG value */
+	max17262_reg_write(priv, HIBCFG, hibcfg);
+
+	/** STEP 3 */
+	/* Read Status register */
+	max17262_reg_read(priv, STATUS, &tmp);
+
+	/* Do not continue until Status.POR bit is cleared */
+	while (tmp & STATUS_POR) {
+		tmp &= STATUS_POR_CLR; /* Clear PowerOnReset bit */
+		max17262_reg_write(priv, STATUS, tmp);
+		k_sleep(K_MSEC(10));
+		max17262_reg_read(priv, STATUS, &tmp);
+	}
```

Comment:
I didn't read the device datasheet, but is this right? Or do you need to call reg_write once and do a loop on reg_read only instead?

Reply:
I re-worked this peace of code. It now only reads the STATUS register, clears the POR bit and write the new value.